README.md

```diff
@@ -0,0 +1,518 @@
+   * [A basic example](#a-basic-example)
+   * [Running doctest for a Cabal package](#running-doctest-for-a-cabal-package)
+* [Writing examples and properties](#writing-examples-and-properties)
+   * [Example groups](#example-groups)
+      * [A note on performance](#a-note-on-performance)
+   * [Setup code](#setup-code)
+   * [Multi-line input](#multi-line-input)
+   * [Multi-line output](#multi-line-output)
+   * [Matching arbitrary output](#matching-arbitrary-output)
+   * [QuickCheck properties](#quickcheck-properties)
+   * [Hiding examples from Haddock](#hiding-examples-from-haddock)
+   * [Using GHC extensions](#using-ghc-extensions)
+* [Limitations](#limitations)
+* [Doctest in the wild](#doctest-in-the-wild)
+* [Development](#development)
+* [Contributors](#contributors)
+
+
+# Getting started
+
+## Installation
+
+`doctest` is available from
+[Hackage](http://hackage.haskell.org/cgi-bin/hackage-scripts/package/doctest).
+Install it with:
+
+    cabal update && cabal install doctest
+
+Make sure that Cabal's `bindir` is on your `PATH`.
+
+On Linux:
+
+    export PATH="$HOME/.cabal/bin:$PATH"
+
+On Mac OS X:
+
+    export PATH="$HOME/Library/Haskell/bin:$PATH"
+
+On Windows:
+
+    set PATH="%AppData%\cabal\bin\;%PATH%"
+
+## A basic example
+
+Below is a small Haskell module.
+The module contains a Haddock comment with some examples of interaction.
+The examples demonstrate how the module is supposed to be used.
+
+```haskell
+-- src/Fib.hs
+module Fib where
+
+-- | Compute Fibonacci numbers
+--
+-- Examples:
+--
+-- >>> fib 10
+-- 55
+--
+-- >>> fib 5
+-- 5
+fib :: Int -> Int
+fib 0 = 0
+fib 1 = 1
+fib n = fib (n - 1) + fib (n - 2)
+
+```
+
+(A comment line starting with `>>>` denotes an _expression_.
+All comment lines following an expression denote the _result_ of that expression.
+Result is defined by what a
+[REPL](http://en.wikipedia.org/wiki/Read-eval-print_loop) (e.g. ghci)
+prints to `stdout` and `stderr` when evaluating that expression.)
+
+With `doctest` you can check whether the implementation satisfies the given
+examples:
+
+```
+doctest src/Fib.hs
+```
+
+
+## Running `doctest` for a Cabal package
+
+The easiest way to run `doctest` for a Cabal package is via `cabal repl --with-ghc=doctest`.
+
+This doesn't make a big difference for a simple package, but in more involved
+situations `cabal` will make sure that all dependencies are available and it
+will pass any required GHC options to `doctest`.
+
+A simple `.cabal` file for `Fib` looks like this:
+
+```cabal
+-- fib.cabal
+cabal-version: 1.12
+
+name:           fib
+version:        0.0.0
+build-type:     Simple
+
+library
+  build-depends: base == 4.*
+  hs-source-dirs: src
+  exposed-modules: Fib
+  default-language: Haskell2010
+
+```
+
+With a `.cabal` file in place, it is possible to run `doctest` via `cabal repl`:
+
+```
+cabal install doctest --overwrite-policy=always && cabal build && cabal repl --build-depends=QuickCheck --with-ghc=doctest
```

Comment:
Rather than installing `doctest`, could it not some be located in the build-tree of the SUT?

Reply:
SUT?

Reply:
https://github.com/sol/doctest-example is an alternative approach. That is add a `doctest` flag to `.cabal` / `package.yaml`:

```yaml
flags:
  doctest:
    manual: True
    default: False

when:
  condition: flag(doctest)
  build-tools: doctest
  dependencies:
    - QuickCheck
    - template-haskell
```
and then invoke `doctest` like so:

```
cabal build -fdoctest && cabal exec -fdoctest -- cabal repl -fdoctest --with-ghc=doctest
```

While this is more robust, it is not easier to document + there will be a bunch of people that will reject the "idea of modifying `.cabal` just for tests", even though, guarded behind a flag, it has no tangible "cost" that I'm aware of.

An other option of course would be to install `doctest` to the current working directory, somewhat cleaner maybe, but you still need to do it before each invocation, assuming that you sometimes switch GHC versions.

I'm not aware of anything that would e.g. tell `cabal exec` to bring some executable into scope via command-line options.  Arguably that is what we would want.

Just to be clear, you don't have to re-install `doctest` before each invocation.  If `doctest` was built with a different version GHC than what you use for e.g. `cabal build` then things will still work:
- `doctest` will run using the version of GHC it was built against
- `cabal repl` may need to rebuild dependencies for that version of GHC, which is fine, as long as it's not `ghc-paths`.  At the point `ghc-paths` gets build via `cabal repl --with-ghc=doctest`, you potentially get into trouble.

Reply:
@andreasabel subject under test, I guess.  Not that I'm aware of.  At least not without modifying `cabal.project[.local]`.  If there is a way, I would be eager to hear about it.

Reply:
I agree I would not want to add such a flag to my projects.
I only know of the `build-tool-depends: doctest:doctest` trick to pass in executable `doctest` to a testsuite.

It might be worthwhile to refer to https://github.com/Hexirp/doctest-driver-gen as well as a way to integrate `doctest`.

Taking a step back, I would want my doctest-suites to run by a simple invocation of `cabal test` or `stack test`, nothing more than that.  This way, they integrate smoothly with both stackage and `haskell-ci`.  Anything more complicated needs manually fiddling all the time, which I find tiresome.

Reply:
@andreasabel the advantage of using `cabal repl` is that it works reliably (modulo `cabal repl` bugs of course, of which there hasn't been a shortage in the past too).  I don't think there is a way to make a cabal test suite work with `Paths_*` (at least I'm not aware of anything).